src/firmware_new/boards/frankie_v1/io/proto_multicast_communication.c

```diff
@@ -0,0 +1,230 @@
+    struct netconn* conn = netconn_new(NETCONN_UDP_IPV6);
+
+    netconn_set_ipv6only(conn, true);
+    netconn_set_recvtimeout(conn, 1000);
+
+    netconn_bind(conn, &comm_profile->multicast_address, comm_profile->port);
+    netconn_join_leave_group(conn, &comm_profile->multicast_address, IP6_ADDR_ANY,
+                             NETCONN_JOIN);
+
+    struct netbuf* rx_buf = NULL;
+    err_t network_err;
+    int protobuf_err;
+
+    for (;;)
+    {
+        network_err = netconn_recv(conn, &rx_buf);
+
+        switch (network_err)
+        {
+            case ERR_TIMEOUT:
+            {
+                io_proto_multicast_communication_profile_notifyEvents(comm_profile,
+                                                                      RECEIVE_TIMEOUT);
+                break;
+            }
+
+            case ERR_OK:
+            {
+                pb_istream_t in_stream = pb_istream_from_buffer(
+                    (uint8_t*)rx_buf->p->payload, rx_buf->p->tot_len);
+
+                // deserialize into buffer
+                io_proto_multicast_communication_profile_acquireLock(comm_profile);
+                protobuf_err = pb_decode(&in_stream, comm_profile->message_fields,
+                                         comm_profile->protobuf_struct);
+                io_proto_multicast_communication_profile_releaseLock(comm_profile);
+
+                // nanopb err logic is inverted, false = error
+                if (protobuf_err)
+                {
+                    io_proto_multicast_communication_profile_notifyEvents(comm_profile,
+                                                                          RECEIVED_PROTO);
+                }
+                break;
+            }
+            default:
+            {
+                netbuf_delete(rx_buf);
+            }
+        }
+    }
+}
+
+void io_proto_multicast_startNetworkingTask(void* arg)
+{
+    MX_LWIP_Init();
+    osEventFlagsSet(networking_event, NETIF_CONFIGURED);
+    osThreadExit();
+}
+
+void io_proto_multicast_communication_profile_acquireLock(
+    ProtoMulticastCommunicationProfile_t* profile)
+{
+    osMutexAcquire(profile->profile_mutex, osWaitForever);
+}
+
+void io_proto_multicast_communication_profile_releaseLock(
+    ProtoMulticastCommunicationProfile_t* profile)
+{
+    osMutexRelease(profile->profile_mutex);
+}
+
+void io_proto_multicast_communication_profile_notifyEvents(
+    ProtoMulticastCommunicationProfile_t* profile, uint32_t events)
+{
+    osEventFlagsSet(profile->communication_event, events);
+}
+
+uint32_t io_proto_multicast_communication_profile_blockUntilEvents(
+    ProtoMulticastCommunicationProfile_t* profile, uint32_t events)
+{
+    return osEventFlagsWait(profile->communication_event, events, osFlagsWaitAny,
+                            osWaitForever);
+}
+
+void io_proto_multicast_communication_profile_destroy(
+    ProtoMulticastCommunicationProfile_t* profile)
+{
+    osMutexDelete(profile->profile_mutex);
+    free(profile);
```

Comment:
```
const char* profile_name
void* protobuf_struct;
const pb_field_t* message_fields
```
all need to be explicitly free'd too?

Reply:
I don't think so because the `profile` does not "own" the `protobuf_struct` other tasks may use it too, `message_fields` are statically defined in a `nanopb` header which do not need to be freed. 

and the char array should get nuked when we free the profile, we don't use malloc to defined the char array, so we don't need to free it